python/ray/train/lightning/lightning_trainer.py

```diff
@@ -0,0 +1,374 @@
+from pytorch_lightning.callbacks import ModelCheckpoint
+from pytorch_lightning.plugins.environments import ClusterEnvironment
+
+import logging
+
+logger = logging.getLogger(__name__)
+
+
+@PublicAPI(stability="alpha")
+class LightningConfigBuilder:
+    """Configuration Class to pass into LightningTrainer."""
+
+    def __init__(self) -> None:
+        """Initialize the configurations with default values."""
+        self._module_class = None
+        self._module_init_config = {}
+        self._trainer_init_config = {}
+        self._trainer_fit_params = {}
+        self._ddp_strategy_config = {}
+        self._model_checkpoint_config = {}
+        self._ray_dataset_iter_config = {}
+
+    def set_module_class(
+        self, module_class: Type[pl.LightningModule]
+    ) -> "LightningConfigBuilder":
+        """Set up the Pytorch Lightning module class.
+
+        Args:
+            module_class: A subclass of ``pytorch_lightning.LightningModule``
+            that defines your model and training logic. Note that this is a
+            class definition instead of a class instance.
+        """
+        if not isclass(module_class):
+            raise ValueError("'module_class' must be a class, not a class instance.")
+        if not issubclass(module_class, pl.LightningModule):
+            raise ValueError(
+                "'module_class' must be a subclass of 'pl.LightningModule'!"
+            )
+        self._module_class = module_class
+        return self
+
+    def set_module_init_config(self, **kwargs) -> "LightningConfigBuilder":
+        """Set up the initialization argument list of your lightning module."""
+        self._module_init_config.update(**kwargs)
+        return self
+
+    def set_trainer_init_config(self, **kwargs) -> "LightningConfigBuilder":
+        """Set up the configurations of ``pytorch_lightning.Trainer``.
+
+        For valid arguments to pass, please refer to:
+        https://pytorch-lightning.readthedocs.io/en/stable/common/trainer.html#init.
+        """
+        self._trainer_init_config.update(**kwargs)
+        return self
+
+    def set_trainer_fit_params(self, **kwargs) -> "LightningConfigBuilder":
+        """Setup the parameter lists for ``pytorch_lightning.Trainer.fit()``.
+
+        For valid arguments to pass, please refer to:
+        https://pytorch-lightning.readthedocs.io/en/stable/common/trainer.html#fit.
+        """
+        self._trainer_fit_params.update(**kwargs)
+        return self
+
+    def set_ddp_strategy_config(self, **kwargs) -> "LightningConfigBuilder":
+        """Set up the configurations of ``pytorch_lightning.Trainer``.
+
+        For valid arguments to pass, please refer to:
+        https://pytorch-lightning.readthedocs.io/en/stable/api/pytorch_lightning.strategies.DDPStrategy.html
+        """
+        self._ddp_strategy_config.update(**kwargs)
+        return self
+
+    def set_model_checkpoint_config(self, **kwargs) -> "LightningConfigBuilder":
```

Comment:
Also, we should mention that calling these `set_...` methods will not replace the full config - it will keep appending on more keys/update existing ones.

I actually like the rllib config naming - just remove the `set_` prefix.

Reply:
also, with a builder, we can group parameters by their semantic meanings, like module_class and module_init_config can be set with a module method:
```
lighting_config = LightningConfigBuilder() \
    .module(cls=ModuleClass, init_config={...}) \
    .trainer(strategy=DDPStrategy(), ...) \
    .model(...) \
    .checkpointing(...) \
    .build()
```
otherwise, we can simply use a @dataclass to replace all the setters.